packages/cursorless-engine/src/languages/TreeSitterQuery/treeSitterQueryCache.ts

```diff
@@ -0,0 +1,72 @@
+import type { Position, TextDocument } from "@cursorless/common";
+import type { QueryMatch } from "./QueryCapture";
+
+export class Cache {
+  private documentUri: string = "";
+  private documentVersion: number = -1;
+  private documentLanguageId: string = "";
+  private startPosition: Position | undefined;
+  private endPosition: Position | undefined;
+  private matches: QueryMatch[] = [];
+
+  isValid(
+    document: TextDocument,
+    startPosition: Position | undefined,
+    endPosition: Position | undefined,
+  ) {
+    return (
+      this.documentUri === document.uri.toString() &&
+      this.documentVersion === document.version &&
```

Comment:
Note to self: this prevents the cache from being racy relying on the document change callback to invalidate it